Load auth manager as a service

henriknyman opened the pull request:
This PR changes how the auth manager is loaded in the edition module. It will load a service `org.neo4j.kernel.api.security.AuthManager.Factory`. If there are more than one service available a specific one can be selected with the configuration value `unsupported.dbms.auth_manager`.
It  also reverses the dependency from the kernel module to the security module, so the kernel api auth interfaces are moved back from the common module.
A separate security module is created in enterprise, that will load an EnterpriseAuthManager. Currently this uses the same implementation as community, but it will be extended with enterprise-specific features.

henriknyman committed:
Load AuthManager as a service

- AuthManager is loaded as a service (kernel.api.security.AuthManager.Factory),
based on the configuration value unsupported.dbms.auth_manager.
- Reversed the dependencies so that security now depends on kernel
- AuthManager now manages the lifecycle of the UserRepository internally
- A separate enterprise auth manager was created

henriknyman committed:
Extract separate interface for UserManager

henriknyman committed:
Change to always require an auth manager

henriknyman committed:
Remove duplicate enterprise auth test

henriknyman commented:
This is currently based upon https://github.com/neo4j/neo4j/pull/6623. The first commit is 	"Moved AuthManager to a kernel extension …	c5eb300"

henriknyman commented:
We should probably squash this a bit since it started out by exploring loading it as a kernel extension.

henriknyman commented:
This is currently based upon https://github.com/neo4j/neo4j/pull/6623. The first commit is  "   Load AuthManager as a service …   3a6cd3f"

benbc commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacadeFactory.java:
```diff
@@ -88,6 +88,7 @@
         // Kept here to have it not be publicly documented.
         public static final Setting<String> lock_manager = setting( "unsupported.dbms.lock_manager", Settings.STRING, "" );
         public static final Setting<String> record_format = setting( "unsupported.dbms.record_format", Settings.STRING, "" );
+        public static final Setting<String> auth_manager = setting( "unsupported.dbms.auth_manager", Settings.STRING, "" );
```
This should be in `GraphDatabaseSettings` and be annotated as `@Internal`, please.

Our other security-related settings live under `dbms.security.*` (or `unsupported.dbms.security.*`) so this should probably be there too.

henriknyman committed:
Move auth_manager config

Move config to GraphDatabaseSettings and rename to
unsupported.dbms.security.auth_manager

henriknyman committed:
Fix failing test

henriknyman committed:
Load auth manager in ha and core-edge

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/EditionModule.java:
```diff
@@ -96,19 +93,37 @@ protected void publishEditionInfo( UsageData sysInfo, DatabaseInfo databaseInfo,
+            {
+                return candidate.newInstance( config, logging.getUserLogProvider() );
+            }
+            else if ( key.equals( "" ) )
```
Could do `key.isEmpty()`

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/EditionModule.java:
```diff
@@ -96,19 +93,37 @@ protected void publishEditionInfo( UsageData sysInfo, DatabaseInfo databaseInfo,
+            }
+        }
+
+        if ( key.equals( "" ) )
```
`key.isEmpty()` is a bit nicer.

Mats-SX commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/EditionModule.java:
```diff
@@ -96,19 +93,37 @@ protected void publishEditionInfo( UsageData sysInfo, DatabaseInfo databaseInfo,
+        if ( key.equals( "" ) )
+        {
+            logging.getUserLog( CommunityFacadeFactory.class )
+                    .error( "No auth manager implementation specified and no default could be loaded." );
```
Perhaps it could be good to output what neo4j thought the default was? To make it easier to understand the error if it happens.

davidegrohmann replied:
It would also nice to explain what the user is supposed to do to make the error go away. E.g., how he can specify an auth manager.

henriknyman replied:
This is based on the same pattern as creating the lock manager. The default config value is the empty string "", which means that it will load the first available service. This is handy, since in community edition this will be the BasicAuthManager and in enterprise it will be the EnterpriseAuthManager. But if you want something other than the default for the edition which you are building you have the possibility to override the decision by setting the config value.
The only difference from what is done with the lock manager is what happens when no service exists at all. We decided that it was unsafe to provide a default with authentication disabled, so instead we abort and throw an exception. 
This is basically an illegal configuration of the neo4j build, so should never happen to a user. It used to be logging the error to the internal log, should I turn it back to that?

henriknyman replied:
I clarified a little bit and added a test

Mats-SX commented on community/security/src/main/java/org/neo4j/server/security/auth/AuthenticationStrategy.java:
```diff
@@ -19,6 +19,8 @@
  */
 package org.neo4j.server.security.auth;
 
+import org.neo4j.kernel.api.security.AuthenticationResult;
```
This import seems astray. Is it necessary to touch this file?

henriknyman replied:
Yes, AuthenticationResult moved from org.neo4j.server.security.auth to org.neo4j.kernel.api.security

Mats-SX replied:
But git says there's nothing removed or changed, just added, implying this file didn't import the class previously, thus not needing to be updated with its move. Is it displaying it wrongly?

Mats-SX commented on community/security/src/main/java/org/neo4j/server/security/auth/BasicAuthSubject.java:
```diff
@@ -22,6 +22,8 @@
 
 import org.neo4j.kernel.api.security.AccessMode;
+import org.neo4j.kernel.api.security.AuthSubject;
+import org.neo4j.kernel.api.security.AuthenticationResult;
```
Again imports added, but no other changes.

henriknyman replied:
Same reason as above

Mats-SX commented on community/security/src/main/java/org/neo4j/server/security/auth/RateLimitedAuthenticationStrategy.java:
```diff
@@ -24,6 +24,8 @@
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicInteger;
 
+import org.neo4j.kernel.api.security.AuthenticationResult;
```
Same as above.

henriknyman replied:
As above :)

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/factory/EditionModule.java:
```diff
@@ -96,19 +93,37 @@ protected void publishEditionInfo( UsageData sysInfo, DatabaseInfo databaseInfo,
+            {
+                logging.getInternalLog( CommunityFacadeFactory.class )
+                        .info( "No auth manager implementation specified, defaulting to '" + candidateId + "'" );
+                return candidate.newInstance( config, logging.getUserLogProvider() );
```
I wonder if failing fast here it would be better...

Mats-SX replied:
You mean not having any default?

henriknyman replied:
Like described above, as with the lock manager, the benefit of loading the first available service is that it is easy to configure the different builds just by adding the right maven dependency. For a production system maybe you want this to be locked down, but it could be done by setting the configuration value.

Mats-SX commented on community/security/src/main/resources/META-INF/services/org.neo4j.kernel.api.security.AuthManager$Factory:
```diff
@@ -0,0 +1 @@
+org.neo4j.server.security.auth.BasicAuthManagerFactory
```
I'm curious, what does this file do?

henriknyman replied:
It specifies a class that provides the service org.neo4j.kernel.api.security.AuthManager.Factory

Mats-SX commented on community/security/src/test/java/org/neo4j/server/security/auth/RateLimitedAuthenticationStrategyTest.java:
```diff
@@ -23,6 +23,7 @@
 
 import java.util.concurrent.TimeUnit;
 
+import org.neo4j.kernel.api.security.AuthenticationResult;
```
Same as above with imports.

henriknyman replied:
... :)

Mats-SX commented on community/server/src/main/java/org/neo4j/server/rest/dbms/UserService.java:
```diff
@@ -48,13 +49,17 @@
-        this.authManager = authManager;
+        if ( !(authManager instanceof UserManager) )
+        {
+            new IllegalArgumentException( "The provided auth manager is not capable of user management" );
```
Missing `throw` here? If yes, a test should probably be added to ensure that this exception behaviour works as intended.

henriknyman replied:
Oops

henriknyman replied:
Fixed

henriknyman committed:
Fix some review comments and add tests

davidegrohmann commented:
reopening against 3.1
